Bluetooth: controller: Extended create connection cancel

cvinayak opened the pull request:
Initial work in progress implementation of Create Connection
Cancel for Extended connection initiation.

Adds implementation to teardown connection initiated at ULL
layer and gracefully release allocated resources.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Extended create connection cancel

Initial work in progress implementation of Create Connection
Cancel for Extended connection initiation.

Adds implementation to teardown connection initiated at ULL
layer and gracefully release allocated resources.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -553,6 +578,46 @@ uint8_t ll_enc_req_send(uint16_t handle, uint8_t const *const rand,
+		conn_lll = scan_coded->lll.conn;
+		LL_ASSERT(conn_lll);
+		scan_coded->lll.conn = NULL;
+
```
Why is there no memq_deinit() and ll_conn_release() for the connection in the Coded PHY scanner? If it is supposed to be the same as for the 1M PHY scanner, then a comment and/or assert would be good.

cvinayak replied:
I will add comments to explain how the single connection context is release considering the 3 cases: 
1. Initiating only on 1M
2. Initiating only on Coded PHY
3. Initiating on both 1M and Coded PHY

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -392,22 +392,47 @@ uint8_t ll_connect_enable(uint8_t is_coded_included)
+		err = 0U;
+	}
+
+	if (!err && scan_coded) {
```
Shouldn't this try to disable the Coded PHY scanner even if there was an error with disabling the 1M PHY scanner?

cvinayak replied:
Until this line, with `ull_scan_disable` if fails, no internal context has been modified. And, as `scan` is non-NULL because it shall be enabled, a failure here indicates a flaw, hence, exiting without further modification to the context of the controller.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -392,22 +392,47 @@ uint8_t ll_connect_enable(uint8_t is_coded_included)
 	}
 
-	conn_lll = scan->lll.conn;
+	conn_lll = scan_lll->conn;
 	if (!conn_lll) {
```
Damn! there is a race here :-(.... will address this as a separate issue.

cvinayak committed:
Bluetooth: controller: Comment on conn context reuse for coded PHY

Added comments to explain the re-use of allocated same
connection context to both 1M and coded PHY scanning context
when both PHY is enabled for initiating connection.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -555,6 +587,57 @@ uint8_t ll_enc_req_send(uint16_t handle, uint8_t const *const rand,
+void ull_master_cleanup(struct node_rx_hdr *rx_free)
+{
+	struct node_rx_ftr *ftr = &rx_free->rx_ftr;
+	struct ll_scan_set *scan =
+		(void *)HDR_LLL2EVT(ftr->param);
```
This could be one line

cvinayak replied:
If this is not too much of a concern, I will stick to 80-column just to avoid github side-by-side compare from not to line wrap.

Thalley replied:
Oh, I guess that might be an issue on smaller screens :) It's a minor comment that doesn't need to be fixed; however you are likely going to see future commits with 100 length lines anyhow ^_^

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -555,6 +587,57 @@ uint8_t ll_enc_req_send(uint16_t handle, uint8_t const *const rand,
+	scan->lll.conn = NULL;
+
+	LL_ASSERT(!conn_lll->link_tx_free);
+	link = memq_deinit(&conn_lll->memq_tx.head,
+			   &conn_lll->memq_tx.tail);
```
This could be one line

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -555,6 +587,57 @@ uint8_t ll_enc_req_send(uint16_t handle, uint8_t const *const rand,
+	/* Determine if coded PHY was also enabled, if so, reset the assigned
+	 * connection context, enabled flag and phy value.
+	 */
+	struct ll_scan_set *scan_coded =
+				ull_scan_is_enabled_get(SCAN_HANDLE_PHY_CODED);
```
This could be one line

Thalley reviewed (approved):
Minor formatting comments, otherwise LGTM
